### PR TITLE
Update i2s.md

### PR DESCRIPTION
**Pull request by stegu**
## What This PR Changes
The documentation was incomplete and vague. I made it better.
This still needs a proper tutorial on how to use buffered writes and reads.

## Contribution Guidelines
- [ ] I confirm that I have read the [contribution guidelines](https://github.com/arduino/docs-content/tree/main/contribution-templates) and comply with them.

### CODE REVIEW

File: content/learn/07.built-in-libraries/02.i2s/i2s.md [hunk 1]
```diff
@@ -29,9 +29,11 @@ The SCK line has a frequency that depends on the sample rate, the number of bits
 
 - **Frequency = SampleRate x BitsPerChannel x numberOfChannels**
 
-In a typical setup, the sender of audio data is called a Transmitter and it transfers data to a Receiver at the other end. The device that controls the bus clock, SCK, together with the Word Select - WS - signal is the Controller in the network and in any network just one device can be Controller at any given time; all the other devices connected are in Peripheral mode. The Controller can be the Transmitter, or the Receiver, or a standalone controller. The digitized audio data sample can have a size ranging from 4 bits up to 32.
+The left and right channels are interleaved in the data stream, and "one sample" is one value for the left channel and one for the right. This means that the FS line oscillates at the sample rate.
```
Comment (marqdevx) [inline content/learn/07.built-in-libraries/02.i2s/i2s.md:32]: I think this is not 100% needed, could you elaborate this paragraph more maybe?

File: content/learn/07.built-in-libraries/02.i2s/i2s.md [hunk 2]
```diff
@@ -77,7 +79,7 @@ nothing
 ### `available()`
 
 #### Description
-Get the number of bytes available for reading from the I2S interface. This is data that has already arrived and was stored in the I2S receive buffer. available() inherits from the Stream utility class.
+Get the number of **bytes** (not the number of samples) available for reading from the I2S interface. This is data that has already arrived and was stored in the I2S receive buffer. available() inherits from the Stream utility class.
```
Comment (marqdevx) [inline content/learn/07.built-in-libraries/02.i2s/i2s.md:82]: Not needed

File: content/learn/07.built-in-libraries/02.i2s/i2s.md [hunk 3]
```diff
@@ -113,31 +115,44 @@ The next sample of incoming I2S data available (or 0 if no data is available)
 ### `write()`
 
 #### Description
-Writes binary data to the I2S interface. This data is sent as a sample or series of samples.
+Writes binary data to the I2S interface. Data is sent either as one sample or a sequence of samples.
+
+The single argument function `write(val)` writes one sample for one channel. A full write requires
+calling this function twice, once for the left channel and once for the right channel.
+If the device buffer is full, the function blocks until the data can be written.
+Before writing, the value `val` will be converted to the appropriate number of bits per sample.
```
Comment (marqdevx) [inline content/learn/07.built-in-libraries/02.i2s/i2s.md:123]: It should be `n` times depending on the amount of channels, if you are only working with one channel then its only once.

File: content/learn/07.built-in-libraries/02.i2s/i2s.md [hunk 4]
```diff
@@ -113,31 +115,44 @@ The next sample of incoming I2S data available (or 0 if no data is available)
...
+
+The two-argument function `write(buf, len)` writes a sequence of samples from the array `buf`.
+Data must be stored in the proper format for the sample size, with the left and right channels
+interleaved, and the argument `len` specifies the number of **bytes** in the buffer, not the
+number of samples. If the device buffer is too full to hold all the data, the function writes
+as much as will fit (possibly nothing), and returns without blocking. The return value is the
+number of bytes that was actually written. It is the responsibility of the application to
+handle any such partial writes and act accordingly.
```
Comment (marqdevx) [inline content/learn/07.built-in-libraries/02.i2s/i2s.md:131]: Be more generic, always referring to the 2 channel scenario

File: content/learn/07.built-in-libraries/02.i2s/i2s.md [hunk 5]
```diff
@@ -113,31 +115,44 @@ The next sample of incoming I2S data available (or 0 if no data is available)
...
+I2S.write(val)  // Might block for a moment until data can be written
+I2S.write(buf, len) // Will not block, but might not write all the data
 
 ```
 
 #### Parameters
 val: a value to send as a single sample
 
-buf: an array to send as a series of samples
+buf: an array to send as a sequence of samples
```
Comment (marqdevx) [inline content/learn/07.built-in-libraries/02.i2s/i2s.md:144]: Not needed

File: content/learn/07.built-in-libraries/02.i2s/i2s.md [hunk 6]
```diff
@@ -113,31 +115,44 @@ The next sample of incoming I2S data available (or 0 if no data is available)
...
 
-len: the length of the buffer
+len: the length of the buffer, in bytes (not in number of samples)
```
Comment (marqdevx) [inline content/learn/07.built-in-libraries/02.i2s/i2s.md:146]: With `in bytes` is enough

File: content/learn/07.built-in-libraries/02.i2s/i2s.md [hunk 7]
```diff
@@ -150,26 +165,67 @@ I2S.availableForWrite()
 none
 
 #### Returns
-the number of bytes available to write.
+size_t
```
Comment (marqdevx) [inline content/learn/07.built-in-libraries/02.i2s/i2s.md:168]: why?

File: content/learn/07.built-in-libraries/02.i2s/i2s.md [hunk 8]
```diff
@@ -150,26 +165,67 @@ I2S.availableForWrite()
...
+- the number of bytes available to write
 
 ### `onTransmit(handler)`
 
 #### Description
-Registers a function to be called when a block of data has been transmitted.
+Registers a function to be called when a block of data has been transmitted. This is used in conjunction with the two-argument function write(buf, len), and the callback function is called when the DMA write operation that was triggered by write(buf, len) finishes. The callback takes no arguments, which mean it cannot determine by itself if the write was completely successful or wrote only some part of the data. Partial writes need to be handled separately by the application.
+
+The callback is not triggered by the single argument version `write(val)`, only by the two argument version `write(buf, len)`. The callback can be registered once at setup of the I2S interface, and will be called for every subsequent write() -- you do not need to register it anew for every write operation.
```
Comment (marqdevx) [inline content/learn/07.built-in-libraries/02.i2s/i2s.md:176]: Could you please send the source of this?

File: content/learn/07.built-in-libraries/02.i2s/i2s.md [hunk 9]
```diff
@@ -150,26 +165,67 @@ I2S.availableForWrite()
...
 #### Returns
 None
 
+### `read()`
+
+#### Description
+Reads binary data from the I2S interface. Data is read either as one sample or a sequence of samples.
+
+The single argument function `read(val)` reads one sample for one channel. A full read requires
+calling this function twice, once for the left channel and once for the right channel.
```
Comment (marqdevx) [inline content/learn/07.built-in-libraries/02.i2s/i2s.md:190]: Be generic

File: content/learn/07.built-in-libraries/02.i2s/i2s.md [hunk 10]
```diff
@@ -150,26 +165,67 @@ I2S.availableForWrite()
...
+If the device buffer is empty, the function blocks until there is data available.
+The return value is always an `int`. You will need to know the number of bits per sample
+to handle the data correctly.
+
+The two-argument function `read(buf, len)` reads a sequence of samples into the array `buf`.
+Data will be stored in the binary format for the current sample size, with the left and right
+channels interleaved. The argument `len` specifies the number of **bytes** in the buffer, not the
```
Comment (marqdevx) [inline content/learn/07.built-in-libraries/02.i2s/i2s.md:197]: Be generic

File: content/learn/07.built-in-libraries/02.i2s/i2s.md [hunk 11]
```diff
@@ -150,26 +165,67 @@ I2S.availableForWrite()
...
+the function retrieves as much as will fit. If the device has less data available, possibly
+nothing, the function will read what is available and return without blocking. The return
+value is the number of bytes that was actually read. It is the responsibility of the
+application to read the return value and act accordingly.
+
+#### Syntax
+
+```
+val = I2S.read()  // Might block until data is available for reading
+I2S.read(buf, len) // Will not block, but might not read all the data
```
Comment (marqdevx) [inline content/learn/07.built-in-libraries/02.i2s/i2s.md:208]:
```suggestion
int val = I2S.read();  // Might block until data is available for reading
int bufferVal = I2S.read(buf, len); // Will not block, but might not read all the data
```

File: content/learn/07.built-in-libraries/02.i2s/i2s.md [hunk 12]
```diff
@@ -150,26 +165,67 @@ I2S.availableForWrite()
...
+
+```
+
+#### Parameters
+
+buf: an array to store a sequence of samples
+
+len: the length of the `buf` array, in bytes (not in number of samples)
```
Comment (marqdevx) [inline content/learn/07.built-in-libraries/02.i2s/i2s.md:216]: Remove the content of the parenthesis

Comment (marqdevx) [inline content/learn/07.built-in-libraries/02.i2s/i2s.md:216]:
```suggestion
len: the length of the `buf` array, in bytes
```

File: content/learn/07.built-in-libraries/02.i2s/i2s.md [hunk 13]
```diff
@@ -150,26 +165,67 @@ I2S.availableForWrite()
...
+
+#### Returns
+size_t
```
Comment (marqdevx) [inline content/learn/07.built-in-libraries/02.i2s/i2s.md:219]: it does return an int, isn't it?

File: content/learn/07.built-in-libraries/02.i2s/i2s.md [hunk 14]
```diff
@@ -150,26 +165,67 @@ I2S.availableForWrite()
...
+
+#### Returns
+size_t 
+- read(buf, len) will return the number of bytes read. Reading that number is essential, as the transmitting device will typically not fill the buffer completely with every read operation.
+
 ### `onReceive(handler)`
 
 #### Description
-Registers a function to be called when a block of data has been received.
+Registers a function to be called when a block of data has been received. This works just like the `onTransmit()` handler, but for the `read(buf, len)` function that receives data from a device. The callback is not triggered for the no-argument version of `read()` that reads a single sample value.
```
Comment (marqdevx) [inline content/learn/07.built-in-libraries/02.i2s/i2s.md:225]: As mentioned before send us the source of this please! 😄

**Review by marqdevx — CHANGES_REQUESTED** [review]
Thanks for the PR!
I have some doubts/suggestions, please if I'm wrong reply to the comment with the reason

Thanks!! 😄

File: content/learn/07.built-in-libraries/02.i2s/i2s.md (hunk 1 above)
Comment (stegu) [reply content/learn/07.built-in-libraries/02.i2s/i2s.md:32]: The comment about the FS line is valid only for 2 channels, so scratch that. The rest is exposition about the I2S protocol, and that can also be scrapped, although documentation on the I2S protocol is generally hard to find and difficult to read. This is a tutorial, not a more narrowly focused "howto", and I like to educate people. (Work related habit, do as you like.)

File: content/learn/07.built-in-libraries/02.i2s/i2s.md (hunk 2 above)
Comment (stegu) [reply content/learn/07.built-in-libraries/02.i2s/i2s.md:82]: On the contrary, my experience is that the most common trap people fall into is to confuse the number of bytes with the number of samples for these functions. This is based on a study of only two test subjects, me and my colleague, but I think it is motivated to point this out. Do as you like.

File: content/learn/07.built-in-libraries/02.i2s/i2s.md (hunk 4 above)
Comment (stegu) [reply content/learn/07.built-in-libraries/02.i2s/i2s.md:131]: Agreed. my bad. See above.

File: content/learn/07.built-in-libraries/02.i2s/i2s.md (hunk 11 above)
Comment (stegu) [reply content/learn/07.built-in-libraries/02.i2s/i2s.md:208]: Yes, of course. Sorry. I'm unfamiliar with the format for this documentation.

File: content/learn/07.built-in-libraries/02.i2s/i2s.md (hunk 12 above)
Comment (stegu) [reply content/learn/07.built-in-libraries/02.i2s/i2s.md:216]: If not mentioned here, definitely deserves pointing out elsewhere. Common pitfall.

Comment (stegu) [reply content/learn/07.built-in-libraries/02.i2s/i2s.md:216]: See above.

File: content/learn/07.built-in-libraries/02.i2s/i2s.md (hunk 14 above)
Comment (stegu) [reply content/learn/07.built-in-libraries/02.i2s/i2s.md:225]: This was verified by my own experiments. I don't see it mentioned anywhere in the docs, and this was a major headache for me. The fact that the loop() function executes at the sample rate would also be worth mentioning. It's not explicitly stated anywhere in the current docs.

**Comment by marqdevx** [comment]
Thanks for the replies, I will check them out. 😄 

Also maybe you should think on creating a tutorial based on your project if you want.

File: content/learn/07.built-in-libraries/02.i2s/i2s.md (hunk 7 above)
Comment (stegu) [reply content/learn/07.built-in-libraries/02.i2s/i2s.md:168]: I'm not sure I get it. Why wouldn't you want to specify the return type for a function in the documentation? Some other pages in the Arduino docs lack this information as well, and I have sometimes been forced to read the source.

File: content/learn/07.built-in-libraries/02.i2s/i2s.md (hunk 8 above)
Comment (stegu) [reply content/learn/07.built-in-libraries/02.i2s/i2s.md:176]: Verified by my own experiments. I had to use a lot of trial and error to get this library to work for me. I didn't find any documentation on this anywhere else, but with the current library code, the callback was most definitely not executed after a single value write when I checked just now.

File: content/learn/07.built-in-libraries/02.i2s/i2s.md (hunk 10 above)
Comment (stegu) [reply content/learn/07.built-in-libraries/02.i2s/i2s.md:197]: Agreed.

**Comment by stegu** [comment]
I will certainly consider writing a "proper" tutorial, although I have a very limited selection of hardware available to me: just the one stereo amplifier and one Arduino board with an on-board microphone. This was just a quick fix for other people to hopefully be able to dodge some of the pitfalls I encountered.

**Comment by marqdevx** [comment]
Sounds good!
Which board are you using btw? 🤔 
I might try it out

**Comment by stegu** [comment]
I am currently experimenting with an Arduino Nano 33 IoT and a 98357 stereo
amplifier. Very low cost and small size, but still very decent sound
quality if you use a decent loudspeaker. My current project is a full
custom Star Wars light saber with purely procedural sound synthesis (no
canned samplings) and a more vividly animated RGB LED strip blade than what
the current commercial controllers offer. It will of course be a lot less
configurable and require reprogramming to change how the thing looks and
sounds and feels. I'm not even going to try to write a user interface for
configuring it.

**Comment by marqdevx** [comment]
Closing the PR due to no changes

Please feel free to re-open it to do the changes

**Comment by stegu** [comment]
Sorry I dropped the ball on this, but I was hit with Covid-1, and I am now
suffering from post-viral brain fatigue. I will need to leave this, and
several other spare time activities, by the wayside until I regain at least
some of my former energy.

**Pull request closed without merging**